### ISSUE TITLE
"ERROR:: Empty mask image" on SPM packs

### ISSUE THREAD

**Issue by cmaumet** [incf-nidash/nidmresults-fslhtml#20]
*This issue was initially raised in https://github.com/incf-nidash/nidmresults-fslhtml/pull/17#issuecomment-361507670 but is moved here as it also affects the current `master` and might be unrelated to changes in #17.*

-----

When running locally:
```
$ python Tests/testError.py 
```

although the tests do pass, I get an `ERROR:: Empty mask image` message multiple times (full log: https://gist.github.com/cmaumet/6eebfb163c2da8af715d7cdecc765dbb).

Because this only happens for SPM packs, I am wondering if there is some bad interaction when using FSL's slicer onto SPM images. My best guess is that there could be NaN's in the SPM images this is causing the issue. 

@TomMaullin: Could you see if we can replace NaN's by zeros (using nibabel) in the images we pass to the slicer command?

**Comment by cmaumet** [comment]
@TomMaullin, I've re-run the tests and then the script your provided as suggested in https://github.com/incf-nidash/nidmresults-fslhtml/pull/17#issuecomment-361564084. Here is the output:

```
$ python errorTest.py 
NIFTI DIMENSION x
ERROR: Could not open file
(b'', b'')

NIFTI DIMENSION y
ERROR: Could not open file
(b'', b'')

NIFTI DIMENSION z
ERROR: Could not open file
(b'', b'')

NIFTI DIMENSION pix
ERROR: Could not open file
(b'', b'')

VAL min
Image Exception : #22 :: ERROR: Could not open image /PATH_TO_SOFTS/nidmresults-fslhtml/Tests/data/ex_spm_default_test/ExcursionSet
(b'', b'')

VAL max
Image Exception : #22 :: ERROR: Could not open image /PATH_TO_SOFTS/nidmresults-fslhtml/Tests/data/ex_spm_default_test/ExcursionSet
(b'', b'')

ERROR: Could not open file
Traceback (most recent call last):
  File "errorTest.py", line 249, in <module>
    template = getTemplate(exc_set)[0]
  File "errorTest.py", line 215, in getTemplate
    if nifDim(exc_set, 'pix') == 1:
  File "errorTest.py", line 36, in nifDim
    dimension = int(float(output[0].decode('utf-8').rstrip('\r|\n').replace(arg, '').replace(' ', '')))
ValueError: could not convert string to float: 
```

I am using FSL 5.0.10. 

Do you get those `"ERROR:: Empty mask image"` on your machine?

**Comment by TomMaullin** [comment]
Hi @cmaumet ,

This seems strange. I can verify it works on my machine. It appears as though on your machine that it is looking for the file:

`/PATH_TO_SOFTS/nidmresults-fslhtml/Tests/data/ex_spm_default_test/ExcursionSet` 

And not:

`/PATH_TO_SOFTS/nidmresults-fslhtml/Tests/data/ex_spm_default_test/ExcursionSet.nii.gz` 

Is this a standard Mac convention? Can you check for me if you have the nifti in that location? And if so can you please now run this test and post the output just so I can verify what file it is looking for:

https://gist.github.com/TomMaullin/b45e1a232aa301c7bf7fa03995e9648d

I also noticed in your log that it seems to have a warning about styling. Could you open a display and let me know if the styling looks correct to you? This will help establish whether it is location the FSL style sheets and nifti templates correctly. (I've also added a test of this into the gist).

Can you also tell me which version of python you are using?

Cheers,
Tom M

**Comment by cmaumet** [comment]
Here is the updated output:

```
$ python errorTest.py 
NIFTI DIMENSION x
/PATH_TO_SOFTS/nidmresults-fslhtml/Tests/data/ex_spm_default_test/ExcursionSet.nii.gz
ERROR: Could not open file
(b'', b'')
/Softs/fsl
```

And after fix (cf. https://gist.github.com/TomMaullin/b45e1a232aa301c7bf7fa03995e9648d#gistcomment-2336620):
```
/PATH_TO_SOFTS/nidmresults-fslhtml (master) $ python errorTest.py 
NIFTI DIMENSION x
/PATH_TO_SOFTS/nidmresults-fslhtml/Tests/data/ex_spm_default_test_err/ExcursionSet.nii.gz
(b'dim1           79\n', b'')
/PATH_TO_SOFTS/fsl
```

(Note: `/PATH_TO_SOFTS`is a placeholder to avoid disclosing locations on my machine).

Styling is fine in the output, but note that the warning is about the file not being closed: `ResourceWarning: unclosed file`.

I am using python 3.5 which is the same as what is used on Travis CI.

**Comment by TomMaullin** [comment]
Hi @cmaumet ,

Ah apologies I wasn't very clear! When I said run the tests I meant 'test_fslviewer.py' not 'testError.py'! This should create the directory you need!

In that case could you please run the original gist (below), either on 'ex_spm_default_test' after running 'test_fslviewer.py' or 'ex_spm_default_test_error' after running 'testError.py':

https://gist.github.com/TomMaullin/b45e1a232aa301c7bf7fa03995e9648d/09dee9e4889d6b547b96d6702a83c348823a5ff2

**Comment by cmaumet** [comment]
Hi @TomMaullin,

Sorry for the delay, here is the updated log:

```
$ python Tests/test_fslViewer.py 
```

```
ERROR:: Empty mask image
ERROR:: Empty mask image
[... Same message again...]
..................
----------------------------------------------------------------------
Ran 18 tests in 0.005s

OK
```

```
$ python errorTest.py 
```
(where `errorTest.py` is the function you provided above)

```
NIFTI DIMENSION x
(b'dim1           79\n', b'')

NIFTI DIMENSION y
(b'dim2           95\n', b'')

NIFTI DIMENSION z
(b'dim3           79\n', b'')

NIFTI DIMENSION pix
(b'pixdim1        2.000000\n', b'')

VAL min
ERROR:: Empty mask image
ERROR:: Empty mask image
ERROR:: Empty mask image
ERROR:: Empty mask image
ERROR:: Empty mask image
ERROR:: Empty mask image
(b'0.000000\n', b'')

VAL max
ERROR:: Empty mask image
ERROR:: Empty mask image
ERROR:: Empty mask image
ERROR:: Empty mask image
ERROR:: Empty mask image
ERROR:: Empty mask image
(b'0.000000\n', b'')

Template
PATH_TO_FSL/data/standard/MNI152_T1_2mm_brain.nii.gz

ScaleFactor
1

Resize
(b'', b'')

Overlay
ERROR:: Empty mask image
ERROR:: Empty mask image
ERROR:: Empty mask image
ERROR:: Empty mask image
ERROR:: Empty mask image
ERROR:: Empty mask image
ERROR:: Empty mask image
ERROR:: Empty mask image
ERROR:: Empty mask image
ERROR:: Empty mask image
ERROR:: Empty mask image
ERROR:: Empty mask image
(b'', b'')

getSliceImageFromNifti
(b'', b'')

```

Does this help?

**Comment by TomMaullin** [comment]
Hi @cmaumet ,

Thank you! Yes this was very helpful! I believe you were right in that the NaN's interfere with how the max and min values of a map are found from an FSL mask on mac for some reason. As a result, the max and min are reported as zeros and the mask ends up empty! I'm currently looking into how to fix this.

**Comment by nicholst** [comment]
Just to record a conversation held with @TomMaullin : We checked the spec, and NIDM Results's  [Excursion Set Map](http://nidm.nidash.org/specs/nidm-results_130.html#section-nidm:'Excursion Set Map') is defined as 

> A map in which the set of elements (e.g., pixels, voxels, vertices, or faces) not selected by the 'Inference' activity is set to zero or NaN.

So we need to accommodate NaN's.

**Comment by cmaumet** [comment]
Hi @TomMaullin! You should be able to use `nibabel` read/write utilities for NIfTIs in combination with [nan_to_num](https://docs.scipy.org/doc/numpy-1.12.0/reference/generated/numpy.nan_to_num.html). Is this enough to get you going?

**Comment by TomMaullin** [comment]
Hi @cmaumet, 

I was just now looking at doing this!... I could possibly get it working using FSL commands instead however, would you rather I used nibabel?

**Comment by cmaumet** [comment]
Yes, as discussed in https://github.com/incf-nidash/nidmresults-fslhtml/issues/13, I think we should use nibabel as much as we can for processings on the images and rely on FSL only for display/formatting functions.

**Comment by TomMaullin** [comment]
This issue was solved in PR #23
